### PR TITLE
Group compatibility imports into a single fixer

### PR DESCRIPTION
**Pull request by UnknownPlatypus**
Hi !

Here is the grouping part discussed in #293.
I measured a **~3% performance improvement** on the same test repository I generally use (5k python files, 483kloc).

I think it's really cheap now to add the older compatibility imports given the new structure. 
Let me know how you feel about it.

### CODE REVIEW

**Review by adamchainz — APPROVED** [review]
Thanks for your patience... my son was born the week of your PR, so I've been somewhat busy since! I think this is a good change, will really reduce toil in adding such replacements, and nice little speed boost. I am changing some of the variable names to make it more obvious, and adding a changelog note, but other than that, no comments. Nice one again @UnknownPlatypus .

File: src/django_upgrade/fixers/compatibility_imports.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            "KeyTransform": "django.db.models.fields.json",
+        },
+    }
+}
+
+
+@lru_cache(maxsize=None)
+def _get_replacements(
+    version: tuple[int, int], looks_like_migrations_file: bool
+) -> Mapping[str, dict[str, str]]:
```
Comment (adamchainz) [inline src/django_upgrade/fixers/compatibility_imports.py:88]: one tweak here: making the return type a `Mapping` allows Mypy to ensure the dict doesn't get mutated outside this function

File: src/django_upgrade/fixers/compatibility_imports.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    if not looks_like_migrations_file:
+        for (
+            target_version,
+            target_replacements,
+        ) in REPLACEMENTS_EXCEPT_MIGRATIONS.items():
+            if target_version <= version:
+                for old_module, rewrite in target_replacements.items():
+                    replacements[old_module].update(rewrite)
+
+    replacements.default_factory = None
```
Comment (adamchainz) [inline src/django_upgrade/fixers/compatibility_imports.py:104]: second tweak: setting `default_factory` to `None` makes this act like a regular dict after this point. always a good idea with `defaultdict`.

Comment (UnknownPlatypus) [reply src/django_upgrade/fixers/compatibility_imports.py:104]: oh nice one! Didn't thought about that.

**Comment by UnknownPlatypus** [comment]
oh absolutely no worries!  
And **congratulation**! I hope everything is going well for you all.

Thanks for the few tweaks and for the changelog, I wasn't sure if you wanted to mention it since it's mostly internal changes.
cheers

**Comment by adamchainz** [comment]
> And **congratulation**! I hope everything is going well for you all.

Thank you, yes it is!

> Thanks for the few tweaks and for the changelog, I wasn't sure if you wanted to mention it since it's mostly internal changes.

I think the performance improvement is worth mentioning :)


See you at Djangocon Europe this year?

**Comment by UnknownPlatypus** [comment]
> See you at Djangocon Europe this year?

Would love to participate to one of these but I'm not sure I'll be available and have the time/budget to travel there. Will you be doing some talks this year ? 

PS: Unrelated but what is you preferred way of contact for a suggestion on one of your blog post ? email ? twitter?

**Comment by adamchainz** [comment]
> Would love to participate to one of these but I'm not sure I'll be available and have the time/budget to travel there. Will you be doing some talks this year ?

I've submitted to give a talk, hopefully I get selected. I'll be there regardless.

> PS: Unrelated but what is your preferred way of contact for a suggestion on one of your blog post ? email ? twitter?

Email preferred but anything works: https://adamj.eu/contact/#details

**Pull request merged**